### PR TITLE
Add UserRegistering event

### PR DESCRIPTION
**Pull request by edalzell**
Other part of https://github.com/statamic/cms/issues/2395.

Based it on how form handles it, added a couple of tests.

Needed for Akismet, which needs to be able to stop user registration the same way it does form submissions.

Closes https://github.com/silentzco/statamic-akismet/issues/5

### CODE REVIEW

File: src/Http/Controllers/UserController.php [hunk 1]
```diff
@@ -70,53 +67,33 @@ public function register(Request $request)
             $user->roles($roles);
         }
 
-        // TODO: Registering event
+        if (UserRegistering::dispatch($user) === false) {
+            return $this->userRegistrationSuccess(true);
+        }
```
Comment (jasonvarga) [inline src/Http/Controllers/UserController.php:72]: If we're aiming to mimic the Form submission behavior (which, you're right, we should) then we should borrow the try/catch style with a SilentFormFailureException from here:

https://github.com/statamic/cms/blob/a3505b787c003cdd3e22db37013305a93391ac4c/src/Http/Controllers/FormController.php#L57-L62

You can leave the existing validation where it is, but we'd just need to let your listener throw a ValidationException as described here: https://statamic.dev/extending/events#formsubmitted

File: src/Http/Controllers/UserController.php [hunk 2]
```diff
@@ -70,53 +67,33 @@ public function register(Request $request)
...
+        $response = request()->has('_redirect') ? redirect(request()->get('_redirect')) : back();
 
-            // If the event returned errors, tack those onto the array.
-            if ($response_errors = array_get($response, 'errors')) {
-                $errors = array_merge($response_errors, $errors);
-                continue;
-            }
-        }
+        session()->flash('user.register.success', __('Registration successful.'));
+        session()->flash('user.registered', ! $silentFailure);
```
Comment (jasonvarga) [inline src/Http/Controllers/UserController.php:95]: We should name this `user.register.user_created` and add an appropriate variable to the tag.

That lives in this trait. I don't love how it's got form-specific stuff in here, but we can leave that refactor for another time. For now, you can duplicate this block and change it to `user_created`.

https://github.com/statamic/cms/blob/a3505b787c003cdd3e22db37013305a93391ac4c/src/Tags/Concerns/GetsFormSession.php#L23-L28

**Review by jasonvarga — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/UserController.php (hunk 1 above)
Comment (edalzell) [reply src/Http/Controllers/UserController.php:72]: Ya I had that the first time I did this, then changed my mind...lol

**Review by jasonvarga — APPROVED** [review]

**Comment by jasonvarga** [comment]
Thanks Erin

**Pull request merged**